mlflow/evaluation.py

```diff
@@ -0,0 +1,424 @@
+        if isinstance(self.data, np.ndarray):
+            return self.data, self.labels
+        elif isinstance(self.data, pd.DataFrame):
+            feature_cols = [x for x in self.data.columns if x != self.labels]
+            return self.data[feature_cols], self.data[self.labels]
+        else:
+            raise ValueError(f'Unsupported data type: {type(self.data)}')
+
+    @staticmethod
+    def _gen_md5_for_arraylike_obj(md5_gen, data):
+        md5_gen.update(pickle.dumps(len(data)))
+        if len(data) < EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH * 2:
+            md5_gen.update(pickle.dumps(data))
+        else:
+            md5_gen.update(pickle.dumps(data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]))
+            md5_gen.update(pickle.dumps(data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]))
+
+    @property
+    def name(self):
+        return self.user_specified_name if self.user_specified_name is not None else self.hash
+
+    @property
+    def hash(self):
+        """
+        Compute a hash from the specified dataset by selecting the first 5 records, last 5 records,
+        dataset size and feeding them through a cheap, low-collision hash function
+        """
+        if self._hash is not None:
+            return self._hash
+        else:
+            md5_gen = hashlib.md5()
+            if isinstance(self.data, np.ndarray):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.labels)
+            elif isinstance(self.data, pd.DataFrame):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                md5_gen.update(self.labels.encode("UTF-8"))
+            return md5_gen.hexdigest()
+
+    @property
+    def metadata(self):
+        metadata = {
+            "hash": self.hash,
+            "path": self.path,
+        }
+        if self.user_specified_name is not None:
+            metadata["name"] = self.user_specified_name
+        return metadata
+
+
+class GetOrCreateRunId:
+    """
+    Get or create a run, return a run_id
+    if user specified a run_id, use it.
+    otherwise if there's an active run, use it
+    otherwise create a managed run.
+    """
+
+    def __init__(self, run_id):
+        self.managed_run_context = None
+        if run_id is not None:
+            self.run_id = run_id
+        elif mlflow.active_run() is not None:
+            self.run_id = mlflow.active_run().info.run_id
+        else:
+            self.run_id = None
+
+    def __enter__(self):
+        if self.run_id is not None:
+            return self.run_id
+        else:
+            self.managed_run_context = mlflow.start_run()
+            return self.managed_run_context.__enter__().info.run_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.managed_run_context is not None:
+            return self.managed_run_context.__exit__(exc_type, exc_val, exc_tb)
```

Comment:
Instead of `GetOrCreateRunId`, could we just use 

`with mlflow.start_run(run_id=run_id)`? From the docs, if `run_id` is `not None`, `start_run()` will activate and resume the specified run. See https://github.com/mlflow/mlflow/blob/9faf3b04c9f6b3766a64d7433dd3e85bcdf27fff/mlflow/tracking/fluent.py#L174-L177.